logstash-core/lib/logstash/util/substitution_variables.rb

```diff
@@ -30,9 +30,17 @@ def deep_replace(value)
   # If value matches the pattern, returns the following precedence : Secret store value, Environment entry value, default value as provided in the pattern
   # If the value does not match the pattern, the 'value' param returns as-is
   def replace_placeholders(value)
-    return value unless value.is_a?(String)
+    case value
```

Comment:
Alternate solution would be to add this to the top, above the existing `String` check, causing us to reenter the function with the `Password` object's value (a `String`) and return the result rewrapped in a new `Password` object. This would eliminate the complexity of tracking klass, and also eliminate the need for `String::new(String)` on inputs that were strings, which would have introduced memory churn.

~~~
if value.kind_of?(::LogStash::Util::Password)
  interpolated = replace_placeholders(value.value)
  return ::LogStash::Util::Password.new(interpolated)
end
~~~

Reply:
@yaauie That is much more elegant, I'll make the change